src/ngraph/node.cpp

```diff
@@ -165,6 +165,29 @@ void Node::set_placement_index(size_t placement)
     m_placement_index = placement;
 }
 
+std::unordered_set<std::string> Node::get_provenance_tags() const
+{
+    return m_provenance_tags;
+}
+
+void Node::add_provenance_tag(const std::string& tag)
+{
+    m_provenance_tags.insert(tag);
+}
+
+size_t Node::remove_provenance_tag(const std::string& tag)
+{
+    return m_provenance_tags.erase(tag);
+}
+
+void Node::merge_provenance_tags_from(std::shared_ptr<const Node> source)
```

Comment:
We usually pass shared pointers by constant reference, i.e. `const std::shared_ptr<const Node>&`.